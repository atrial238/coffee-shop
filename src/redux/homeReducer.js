import {getBestCoffeeAPI} from '../API/api';

const setIsError = (bool) => ({type: types.SET_IS_ERROR, body: bool}),
		setIsLoading = (bool) => ({type: types.SET_IS_LOADING, body: bool}),
		setBestCoffee = (data) => ({type: types.GET_BEST_COFFEE, body: data});

export const types = {
	GET_BEST_COFFEE: 'home_reducer/GET_BEST_COFFEE',
	SET_IS_LOADING: 'home_reducer/SET_IS_LOADING',
	SET_IS_ERROR: 'home_reducer/SET_IS_ERROR',
}

export const getBestCoffee = (methodAPI = getBestCoffeeAPI) => (dispatch) =>  {
	dispatch(setIsError(false))
	dispatch(setIsLoading(true))
	return methodAPI()
		.then(res => {
			dispatch(setIsLoading(false))
			dispatch(setBestCoffee(res.bestCoffee))
		})
		.catch(() => {
			dispatch(setIsLoading(false))
			dispatch(setIsError(true))
		})
}

const initState = {
	bestCoffe: [],
	isLoading: false,
	isError: false
}

export const homeReducer = (state = initState, action) => {
	switch(action.type){
		case types.GET_BEST_COFFEE:
			return {...state, bestCoffe: action.body}
		case types.SET_IS_LOADING:
			return {...state, isLoading: action.body}
		case types.SET_IS_ERROR:
			return {...state, isError: action.body}
		default: 
			return state;
	}
}