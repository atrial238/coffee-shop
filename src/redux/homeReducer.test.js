import { getBestCoffee, homeReducer, types } from './homeReducer';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

const mockServiceCreator = (body, succeeds = true) => () =>
	new Promise((resolve, reject) => {
		setTimeout(() => (succeeds ? resolve(body) : reject('error')), 10);
});

describe('homeReducer', () => {
	const initState = {
		bestCoffe: [],
		isLoading: false,
		isError: false
	}

	test('shoud return initial state', () => {
		const initialState = homeReducer(undefined, {});
		expect(initialState).toMatchSnapshot();
	});

	test('shoud set bestCoffee array', () => {
		const action =  {type: types.GET_BEST_COFFEE, body: [{coffeeone: 'one'}, {coffeeTwo: 'two'}]};
		const state = homeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set isLoading true', () => {
		const action =  {type: types.SET_IS_LOADING, body: true};
		const state = homeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})

	test('shoud set isError true', () => {
		const action = {type: types.SET_IS_ERROR, body: true};
		const state = homeReducer(initState, action);
		expect(state).toMatchSnapshot();
	})
})

describe('getBestCoffee thunk', () => {
	let store;

	beforeEach(() => store = mockStore({}));
 
	it('dispatch action to store when getBestCoffee success', () => {
	
	const expectedActions = [
		{type: types.SET_IS_ERROR, body: false},
		{type: types.SET_IS_LOADING, body: true},
		{type: types.SET_IS_LOADING, body: false},
		{type: types.GET_BEST_COFFEE, body: 'data'}
	]
	
	return store.dispatch(getBestCoffee(mockServiceCreator({bestCoffee: 'data'})))
		.then(() => expect(store.getActions()).toEqual(expectedActions));
	});

	it('dispatch action to store when getBestCoffee failed', () => {
	
		const expectedActions = [
			{type: types.SET_IS_ERROR, body: false},
			{type: types.SET_IS_LOADING, body: true},
			{type: types.SET_IS_LOADING, body: false},
			{type: types.SET_IS_ERROR, body: true},
		]
		
		return store.dispatch(getBestCoffee(mockServiceCreator(null, false)))
			.then(() => expect(store.getActions()).toEqual(expectedActions))
			.catch(() => expect(store.getActions()).toEqual(expectedActions));
	});
 })