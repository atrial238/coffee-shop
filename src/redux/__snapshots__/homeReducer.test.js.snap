// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`homeReducer shoud add new bestCoffee array 1`] = `
Object {
  "bestCoffe": Array [
    Object {
      "coffeeone": "one",
    },
    Object {
      "coffeeTwo": "two",
    },
  ],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud return initial state 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set bestCoffee array 1`] = `
Object {
  "bestCoffe": Array [
    Object {
      "coffeeone": "one",
    },
    Object {
      "coffeeTwo": "two",
    },
  ],
  "isError": false,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set isError true 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": true,
  "isLoading": false,
}
`;

exports[`homeReducer shoud set isLoading true 1`] = `
Object {
  "bestCoffe": Array [],
  "isError": false,
  "isLoading": true,
}
`;

exports[`thunk  success 1`] = `
Array [
  Object {
    "body": false,
    "type": "home_reducer/SET_IS_ERROR",
  },
  Object {
    "body": true,
    "type": "home_reducer/SET_IS_LOADING",
  },
  Object {
    "body": false,
    "type": "home_reducer/SET_IS_LOADING",
  },
  Object {
    "body": false,
    "type": "home_reducer/SET_IS_ERROR",
  },
  Object {
    "body": Array [
      Object {
        "country": "Brazil",
        "id": 0,
        "imgSrc": "1.jpg",
        "nameCoffee": "Solimo",
        "price": 10.37,
        "weight": 1,
      },
      Object {
        "country": "Columbia",
        "id": 1,
        "imgSrc": "2.jpg",
        "nameCoffee": "Presto",
        "price": 15.99,
        "weight": 1,
      },
      Object {
        "country": "Brazil",
        "id": 3,
        "imgSrc": "3.jpg",
        "nameCoffee": "Aromistico",
        "price": 6.99,
        "weight": 1,
      },
    ],
    "type": "home_reducer/GET_BEST_COFFEE",
  },
]
`;

exports[`thunk create 1`] = `
Array [
  Object {
    "body": false,
    "type": "home_reducer/SET_IS_ERROR",
  },
  Object {
    "body": true,
    "type": "home_reducer/SET_IS_LOADING",
  },
  Object {
    "body": false,
    "type": "home_reducer/SET_IS_LOADING",
  },
  Object {
    "body": false,
    "type": "home_reducer/SET_IS_ERROR",
  },
  Object {
    "body": Array [
      Object {
        "country": "Brazil",
        "id": 0,
        "imgSrc": "1.jpg",
        "nameCoffee": "Solimo",
        "price": 10.37,
        "weight": 1,
      },
      Object {
        "country": "Columbia",
        "id": 1,
        "imgSrc": "2.jpg",
        "nameCoffee": "Presto",
        "price": 15.99,
        "weight": 1,
      },
      Object {
        "country": "Brazil",
        "id": 3,
        "imgSrc": "3.jpg",
        "nameCoffee": "Aromistico",
        "price": 6.99,
        "weight": 1,
      },
    ],
    "type": "home_reducer/GET_BEST_COFFEE",
  },
]
`;
